traffic_ops/traffic_ops_golang/deliveryservice/deliveryservices.go

```diff
@@ -699,62 +759,301 @@ WHERE
+	newHostName := ""
+	newCDNName := ""
+	if dsType.HasSSLKeys() {
+		newCDNName, err = getCDNName(*dsNullableV30.CDNID, tx)
+		newHostName, err = getHostName(dsNullableV30.Protocol, *dsNullableV30.Type, *dsNullableV30.RoutingName, *dsNullableV30.MatchList, cdnDomain)
+		if err != nil {
+			return nil, http.StatusInternalServerError, nil, errors.New("getting cdnname/hostname after update: " + err.Error())
+		}
+	}
+
+	if newDSType.HasSSLKeys() && (oldHostName != newHostName || oldCDNName != newCDNName || oldRoutingName != *dsNullableV30.RoutingName) {
+		return nil, http.StatusForbidden, nil, errors.New("delivery service has ssl keys that cannot be automatically changed")
+	}
+
+	if err := EnsureParams(tx, *dsNullableV30.ID, *dsNullableV30.XMLID, dsNullableV30.EdgeHeaderRewrite, dsNullableV30.MidHeaderRewrite, dsNullableV30.RegexRemap, dsNullableV30.CacheURL, dsNullableV30.SigningAlgorithm, newDSType, dsNullableV30.MaxOriginConnections); err != nil {
+		return nil, http.StatusInternalServerError, nil, errors.New("ensuring dsNullableV30 parameters:: " + err.Error())
+	}
+
+	if err := updatePrimaryOrigin(tx, user, dsNullableV30); err != nil {
+		return nil, http.StatusInternalServerError, nil, errors.New("updating delivery service: " + err.Error())
+	}
+
+	dsNullableV30.LastUpdated = &lastUpdated
+
+	// the update may change or delete the query params -- delete existing and re-add if any provided
+	q := `DELETE FROM deliveryservice_consistent_hash_query_param WHERE deliveryservice_id = $1`
+	if res, err := tx.Exec(q, *dsNullableV30.ID); err != nil {
+		return nil, http.StatusInternalServerError, nil, fmt.Errorf("deleting consistent hash query params for dsNullableV30 %s: %s", *dsNullableV30.XMLID, err.Error())
+	} else if c, _ := res.RowsAffected(); c > 0 {
+		api.CreateChangeLogRawTx(api.ApiChange, "DS: "+*dsNullableV30.XMLID+", ID: "+strconv.Itoa(*dsNullableV30.ID)+", ACTION: Deleted "+strconv.FormatInt(c, 10)+" consistent hash query params", user, tx)
+	}
+
+	if c, err := createConsistentHashQueryParams(tx, *dsNullableV30.ID, dsNullableV30.ConsistentHashQueryParams); err != nil {
+		usrErr, sysErr, code := api.ParseDBError(err)
+		return nil, code, usrErr, sysErr
+	} else {
+		api.CreateChangeLogRawTx(api.ApiChange, "DS: "+*dsNullableV30.XMLID+", ID: "+strconv.Itoa(*dsNullableV30.ID)+", ACTION: Created "+strconv.Itoa(c)+" consistent hash query params", user, tx)
+	}
+
+	if err := api.CreateChangeLogRawErr(api.ApiChange, "Updated dsNullableV30: "+*dsNullableV30.XMLID+" id: "+strconv.Itoa(*dsNullableV30.ID), user, tx); err != nil {
+		return nil, http.StatusInternalServerError, nil, errors.New("writing change log entry: " + err.Error())
+	}
+	ds = (*tc.DeliveryServiceV31)(&dsNullableV30)
+	return ds, http.StatusOK, nil, nil
+}
+
+func updateV30(w http.ResponseWriter, r *http.Request, inf *api.APIInfo, ds *tc.DeliveryServiceV30) (*tc.DeliveryServiceV30, int, error, error) {
```

Comment:
Per my earlier comment above, this function should look just like `createV30`, except with the DB query that reads the existing value of the new field.